File: rust/src/websocket/subscriptions.rs
use serde::{Deserialize, Serialize};

use super::client::TradingClient;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelConfig {
    pub name: String,
    pub symbols: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubscriptionRequest {
    pub action: String,
    pub channels: Vec<ChannelConfig>,
}

impl TradingClient {
    pub async fn subscribe_channel(&self, channel_name: &str, symbols: Vec<String>) -> Result<(), Box<dyn std::error::Error>> {
        let req = SubscriptionRequest {
            action: "subscribe".to_string(),
            channels: vec![ChannelConfig {
                name: channel_name.to_string(),
                symbols,
            }],
        };

        self.send_raw(req).await
    }

    pub async fn unsubscribe_channel(&self, channel_name: &str, symbols: Vec<String>) -> Result<(), Box<dyn std::error::Error>> {
        let req = SubscriptionRequest {
            action: "unsubscribe".to_string(),
            channels: vec![ChannelConfig {
                name: channel_name.to_string(),
                symbols,
            }],
        };

        self.send_raw(req).await
    }

    pub async fn subscribe_trades(&self, symbols: Vec<String>, board_id: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
        let boards = match board_id {
            Some(b) => vec![b.to_string()],
            None => vec!["G1", "G3", "G4", "G7", "T1", "T2", "T3", "T4", "T6"].into_iter().map(String::from).collect(),
        };

        for board in boards {
            let channel = format!("tick.{}.{}", board, self.config.encoding);
            self.subscribe_channel(&channel, symbols.clone()).await?;
        }
        Ok(())
    }

    pub async fn subscribe_trade_extra(&self, symbols: Vec<String>, board_id: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
        let boards = match board_id {
            Some(b) => vec![b.to_string()],
            None => vec!["G1", "G3", "G4", "G7", "T1", "T2", "T3", "T4", "T6"].into_iter().map(String::from).collect(),
        };

        for board in boards {
            let channel = format!("tick_extra.{}.{}", board, self.config.encoding);
            self.subscribe_channel(&channel, symbols.clone()).await?;
        }
        Ok(())
    }


    pub async fn subscribe_ohlc(&self, symbols: Vec<String>, resolution: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
        let resolutions = match resolution {
            Some(r) => vec![r.to_string()],
            None => vec!["1", "3", "5", "15", "30", "1H", "1D", "1W"].into_iter().map(String::from).collect(),
        };

        for res in resolutions {
            let channel = format!("ohlc.{}.{}", res, self.config.encoding);
            self.subscribe_channel(&channel, symbols.clone()).await?;
        }
        Ok(())
    }

    pub async fn subscribe_ohlc_closed(&self, symbols: Vec<String>, resolution: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
        let resolutions = match resolution {
            Some(r) => vec![r.to_string()],
            None => vec!["1", "3", "5", "15", "30", "1H", "1D", "1W"].into_iter().map(String::from).collect(),
        };

        for res in resolutions {
            let channel = format!("ohlc_closed.{}.{}", res, self.config.encoding);
            self.subscribe_channel(&channel, symbols.clone()).await?;
        }
        Ok(())
    }

    pub async fn subscribe_quotes(&self, symbols: Vec<String>, board_id: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
        let boards = match board_id {
            Some(b) => vec![b.to_string()],
            None => vec!["G1", "G2", "G3", "G4", "G5", "G6", "G7"].into_iter().map(String::from).collect(),
        };

        for board in boards {
            let channel = format!("top_price.{}.{}", board, self.config.encoding);
            self.subscribe_channel(&channel, symbols.clone()).await?;
        }
        Ok(())
    }

    pub async fn subscribe_orders(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.subscribe_channel("orders", vec![]).await
    }

    pub async fn subscribe_positions(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.subscribe_channel("positions", vec![]).await
    }

    pub async fn subscribe_account(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.subscribe_channel("account", vec![]).await
    }
}

File: go/websocket/subscriptions.go
package websocket

import "fmt"

type ChannelConfig struct {
	Name    string   `json:"name" msgpack:"name"`
	Symbols []string `json:"symbols" msgpack:"symbols"`
}

type SubscriptionRequest struct {
	Action   string          `json:"action" msgpack:"action"`
	Channels []ChannelConfig `json:"channels" msgpack:"channels"`
}

var defaultBoards = []string{"G1", "G3", "G4", "G7", "T1", "T2", "T3", "T4", "T6"}

func (c *TradingClient) subscribeChannel(channelName string, symbols []string) error {
	if !c.isAuthenticated {
		return fmt.Errorf("must authenticate before subscribing")
	}

	req := SubscriptionRequest{
		Action: "subscribe",
		Channels: []ChannelConfig{
			{Name: channelName, Symbols: symbols},
		},
	}

	// Store for reconnection logic
	c.subsMutex.Lock()
	c.subscriptions[channelName] = req
	c.subsMutex.Unlock()

	return c.sendMessage(req)
}

func (c *TradingClient) SubscribeTrades(symbols []string, boardID string) error {
	boards := defaultBoards
	if boardID != "" {
		boards = []string{boardID}
	}
	for _, board := range boards {
		channel := fmt.Sprintf("tick.%s.%s", board, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeTradeExtra(symbols []string, boardID string) error {
	boards := defaultBoards
	if boardID != "" {
		boards = []string{boardID}
	}
	for _, board := range boards {
		channel := fmt.Sprintf("tick_extra.%s.%s", board, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeExpectedPrice(symbols []string, boardID string) error {
	boards := defaultBoards
	if boardID != "" {
		boards = []string{boardID}
	}
	for _, board := range boards {
		channel := fmt.Sprintf("expected_price.%s.%s", board, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeForeignInvestor(symbols []string, boardID string) error {
	boards := defaultBoards
	if boardID != "" {
		boards = []string{boardID}
	}
	for _, board := range boards {
		channel := fmt.Sprintf("foreign_investor.%s.%s", board, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeQuotes(symbols []string, boardID string) error {
	boards := []string{"G1", "G2", "G3", "G4", "G5", "G6", "G7"}
	if boardID != "" {
		boards = []string{boardID}
	}
	for _, board := range boards {
		channel := fmt.Sprintf("top_price.%s.%s", board, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeOHLC(symbols []string, resolution string) error {
	resolutions := []string{"1", "3", "5", "15", "30", "1H", "1D", "1W"}
	if resolution != "" {
		resolutions = []string{resolution}
	}

	for _, res := range resolutions {
		channel := fmt.Sprintf("ohlc.%s.%s", res, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeOHLCClosed(symbols []string, resolution string) error {
	resolutions := []string{"1", "3", "5", "15", "30", "1H", "1D", "1W"}
	if resolution != "" {
		resolutions = []string{resolution}
	}

	for _, res := range resolutions {
		channel := fmt.Sprintf("ohlc_closed.%s.%s", res, c.config.Encoding)
		if err := c.subscribeChannel(channel, symbols); err != nil {
			return err
		}
	}
	return nil
}

func (c *TradingClient) SubscribeMarketIndex(marketIndex string) error {
	channel := fmt.Sprintf("market_index.%s.%s", marketIndex, c.config.Encoding)
	return c.subscribeChannel(channel, []string{})
}

func (c *TradingClient) SubscribeOrders() error {
	return c.subscribeChannel("orders", []string{})
}

func (c *TradingClient) SubscribePositions() error {
	return c.subscribeChannel("positions", []string{})
}

func (c *TradingClient) SubscribeAccount() error {
	return c.subscribeChannel("account", []string{})
}

func (c *TradingClient) Unsubscribe(channelName string, symbols []string) error {
	req := SubscriptionRequest{
		Action: "unsubscribe",
		Channels: []ChannelConfig{
			{Name: channelName, Symbols: symbols},
		},
	}

	c.subsMutex.Lock()
	delete(c.subscriptions, channelName) // Simplistic map update logic
	c.subsMutex.Unlock()

	return c.sendMessage(req)
}
